streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -493,7 +531,9 @@ boolean tryToCompleteRestoration(final long now, final java.util.function.Consum
         final List<Task> activeTasks = new LinkedList<>();
         for (final Task task : tasks.allTasks()) {
             try {
-                task.initializeIfNeeded();
+                if (task.initializeIfNeeded() && stateUpdater != null) {
+                    stateUpdater.add(task);
+                }
```

Comment:
The motivation is that, if we add the newly created tasks to the state updater immediately, they will fail since the state updater expect their states to be `restoring / running`, i.e. after the tasks have already initialized.

Reply:
We discussed this offline yesterday. For future reference, even if we initialize the tasks in `createNewTasks()` and we do not run into failures in the state updater, we are afraid that the initialization would take too long due to state store initializations. The handle assignment logic is executed during a rebalance, i.e., within a call to `Consumer#poll()` and it should not take too long to avoid timeouts.